### ISSUE TITLE
Metric: DSS

### ISSUE THREAD

**Issue by leihuayi** [photosynthesis-team/piq#189]
**Is your feature request related to a problem? Please describe.**
DCT Subbands Similarity exploits important characteristics of human visual perception by measuring change in structural information in subbands in the discrete cosine transform (DCT) domain and weighting the quality estimates for these subbands.
*Performance* : Works well according to the following studies:
- [A comprehensive evaluation of full-reference image quality assessment algorithms on KADID-10k](https://arxiv.org/pdf/1907.02096), 2019
- [A Comparative Study of Image Quality Assessment Models through Perceptual Optimization](A Comparative Study of Image Quality Assessment Models through Perceptual Optimization), 2020, 2020

**Describe the solution you'd like**
A pytorch implementation of DSS

**Describe alternatives you've considered**
I have coded a python version out of the matlab code (results exactly the same as matlab). However I feel it would be more useful to the community if it were part of your repo which is very complete and functional.
I am not very knowledgeable in Pytorch but if you do not have time I can try to implement this feature as contributor.

**Additional context**
Original paper : http://sipl.eelabs.technion.ac.il/wp-content/uploads/sites/6/2016/09/paper15-Image-Quality-Assessment-Based-on-DCT-Subband-Similarity.pdf
Matlab code : https://fr.mathworks.com/matlabcentral/fileexchange/53708-dct-subband-similarity-index-for-measuring-image-quality
Python code : https://github.com/leihuayi/FRIQAMetrics/blob/master/friqametrics/dss.py

**Comment by zakajd** [comment]
Hi!
Thanks for pointing out those metrics. I've also seen them during paper review, but initially decided not to add because of high work load with other metrics. 

It would be nice if you port your code to PyTorch and open PR. Same for SR-SIM (#188).

**Comment by leihuayi** [comment]
Understood I'll do that in the next weeks. Thank @zakajd

**Comment by zakajd** [comment]
Hi @leihuayi 
Have you made any progress with DSS port to PyTorch?
We have students from Russia willing to help with library development, so I'll assign one of them to this issue if you didn't started yet 😄

**Comment by AliyaDav** [comment]
@zakajd, I can proceed with implementation in case nobody is working on it at the moment

**Comment by leihuayi** [comment]
@zakajd @AliyaDav I have finished a first version which outputs a result, but I haven't tested yet if it gives exactly the same results at matlab. It's on branch feature/DSS. I can try to finish it in the next days.

**Comment by AliyaDav** [comment]
@leihuayi I will be happy to contribute. I can run the test to see if the results correspond to the those in the original paper and then implement PyTroch version.

**Comment by leihuayi** [comment]
@AliyaDav Pytorch version is already implemented in branch feature/DSS in my fork, now it's more a matter of tweaking a bit the code to make it fit perfectly matlab results.
If you want to run some tests on your side, here are some matlab results I ran with latest version of matlab online (they might differ slightly from octave but our goal here is to fit matlab):

- DSS on (goldhill.gif, goldhill_jpeg.gif): 0.452171811650662
- DSS on (rgb2gray(i01.bmp), rbg2gray(i01_01_5.bmp)): 0.771774362587157

**Comment by leihuayi** [comment]
Hello,
I ran some tests and I am stuck in a situation I cannot disentangle.
It seems piq's `gaussian_filter`  form filters utility and matlab's `fspecial('gaussian')` slightly differ, and torch's `F.conv2d` differes a bit from matlab's `imfilter`. 
Using only piq and pytorch elements, I tried many things like changing the padding, but after many trials I didn't succeed to make the difference between matlab result and pytorch result decrease below 10e-3 / 10e-4.

If in `_subband_similarity` code we replace this piece of code:
```
    kernel = gaussian_filter(kernel_size=kernel_size, sigma=sigma)
    kernel = kernel.view(1, 1, kernel_size, kernel_size).to(x)
    mu_x = F.conv2d(x, kernel, padding=kernel_size//2)
    mu_y = F.conv2d(y, kernel, padding=kernel_size//2)

    sigma_xx = F.conv2d(x * x, kernel, padding=kernel_size//2) - mu_x ** 2
    sigma_yy = F.conv2d(y * y, kernel, padding=kernel_size//2) - mu_y ** 2
    sigma_xy = F.conv2d(x * y, kernel, padding=kernel_size//2) - mu_x * mu_y
```

by this :
    window = matlab_style_gauss2D(shape=(kernel_size,kernel_size), sigma=sigma)
    x = x.numpy()[0,0,:,:]
    y = y.numpy()[0,0,:,:]
    mu1 = ndimage.convolve(x, window, mode="constant")
    mu2 = ndimage.convolve(y, window, mode="constant")
    sigma1_sq = ndimage.convolve(x*x, window, mode="constant") - mu1*mu1
    sigma2_sq = ndimage.convolve(y*y, window, mode="constant") - mu2*mu2
    sigma12 = ndimage.convolve(x*y, window, mode="constant") - mu1*mu2

    sigma_xx = torch.tensor(sigma1_sq).view(1,1,sigma1_sq.shape[-2], sigma1_sq.shape[-1])
    sigma_yy = torch.tensor(sigma2_sq).view(1,1,sigma2_sq.shape[-2], sigma2_sq.shape[-1])
    sigma_xy = torch.tensor(sigma12).view(1,1,sigma12.shape[-2], sigma12.shape[-1])
   ```
   
where ndimage is scipy's ndimage module and:
```
def matlab_style_gauss2D(shape=(3,3),sigma=0.5):
    """
    2D gaussian mask - should give the same result as MATLAB's
    fspecial('gaussian',[shape],[sigma])
    """
    m,n = [(ss-1.)/2. for ss in shape]
    y,x = np.ogrid[-m:m+1,-n:n+1]
    h = np.exp( -(x*x + y*y) / (2.*sigma*sigma) )
    h[ h < np.finfo(h.dtype).eps*h.max() ] = 0
    sumh = h.sum()
    if sumh != 0:
        h /= sumh
    return h
```

then we can reduce more the difference in some cases, like reach DSS(goldhill.gif, goldhill_jpeg.gif) = 0.45217577.

Here are the different results for some pairs of images if you want to judge if the difference is problematic:

- goldhill.gif : matlab: 0.452171811650662, pytorch: 0.45290700
- i01.bmp : matlab: 0.771774362587157, pytorch: 0.77168512
- 228.jpg : matlab: 0.909963573175721, pytorch: 0.90969026
- 12.jpg : matlab: 0.891021423992554, pytorch: 0.89124078
- High_g1 : matlab: 0.923184009672770, pytorch: 0.92314106
- Low_a1 : matlab: 0.848313138990592, pytorch: 0.84849083
- Low_a3: matlab: 0.765425951110176, pytorch: 0.76549941
